src/Authentication/Authentication/test/Find-MgGraphPermission.Tests.ps1

```diff
@@ -2,15 +2,170 @@
 Describe "the Find-MgGraphPermission Command" {
-    BeforeAll{
+
+    BeforeAll {
         . (join-path $PSScriptRoot  ..\custom\Find-MgGraphPermission.ps1)
+        . (Join-Path $PSScriptRoot  .\Find-MgGraphPermissionTestfile.ps1)
+    }
+
+    Context "When executing the command with empty service principal results from MS Graph" {
+        BeforeAll {
+            Mock Invoke-MgGraphRequest {
+
+                [PSCustomObject] @{
+                    Value = [PSCustomObject] @{
+                        Oauth2PermissionScopes = @()
+                        AppRoles = @()
+                    }
+                }
+            }
+        }
+
+        It 'Executes successfully with no parameters' {
+            { Find-MgGraphPermission | Out-Null } | Should -Not -Throw
+            Assert-MockCalled Invoke-MgGraphRequest
+        }
     }
 
-    Context "when executing the command" {
+    Context "When executing the command using a constrained set of permissions returned by MS Graph" {
+        BeforeAll {
+            Mock Invoke-MgGraphRequest {$permissionData}
+        }
+
+        It 'Executes successfully with no parameters' {
+            { Find-MgGraphPermission | Out-Null } | Should -Not -Throw
+            Assert-MockCalled Invoke-MgGraphRequest
+        }
+
+        It "Retrieves the expected set of delegated and app-only permissions when a search string is specified" {
+            { Find-MgGraphPermission mail | Out-Null } | Should -Not -Throw
+            Assert-MockCalled Invoke-MgGraphRequest
+        }
+
+        It "Returns nothing and throws no exception if a search string is specified and there is no match" {
+            { Find-MgGraphPermission 'Microsoft is the greatest company' | Out-Null } | Should -Not -Throw
+            Assert-MockCalled Invoke-MgGraphRequest
+        }
+    }
+
+    Context "When executing the command without a connection to MS Graph" {
+        BeforeAll {
+            Mock Invoke-MgGraphRequest {
+
+                Throw [System.Management.Automation.ValidationMetadataException]::new('mock connection error message')
+
+            }
+        }
+
+        It 'Executes successfully with no parameters' {
+            { Find-MgGraphPermission | Out-Null } | Should -Not -Throw
+            Assert-MockCalled Invoke-MgGraphRequest
+        }
+
+        It "Retrieves the expected set of delegated and app-only permissions when a search string is specified" {
+            { Find-MgGraphPermission mail | Out-Null } | Should -Not -Throw
+            Assert-MockCalled Invoke-MgGraphRequest
+        }
+
+        It "Returns nothing and throws no exception if a search string is specified and there is no match" {
+            { Find-MgGraphPermission 'Microsoft is the greatest company' | Out-Null } | Should -Not -Throw
+            Assert-MockCalled Invoke-MgGraphRequest
+        }
+    }
+
+    Context "When executing the command after initially not having a connection to MS Graph but getting it later" {
+
```

Comment:
One more thing -- in these tests you'll need some sort of state that tells you whether you are using the mocked service principal data or the static file from the source code. An easy way to check would be for your mock of `Invoke-MgGraphRequest` to also set some external state (e.g. a script variable) that you reset before you invoke `Find-MgGraphRequest`. If that mock changes it, you know for sure your second mock got called for instance. We may need to add deeper logic to test state variables of the code itself to see that they are also set correctly.